### PR TITLE
Use `dlopen` to load NCCL.

### PR DESCRIPTION
**Pull request by trivialfis**
This PR adds optional support for loading nccl with `dlopen` as an alternative of compile time linking. This is to address the size bloat issue with the PyPI binary release.
- Add CMake option to load `nccl` at runtime.
- Add an NCCL stub.

After this, `nccl` will be fetched from PyPI when using pip to install XGBoost, either by user or by `pyproject.toml`. Others who want to link the nccl at compile time can continue to do so without any change.

At the moment, this is Linux only since we only support MNMG on Linux.

### CODE REVIEW

**Comment by trivialfis** [comment]
cc @rongou @hcho3 @RAMitchell @jakirkham This is ready for review now.

**Comment by trivialfis** [comment]
cc @WeichenXu123 .

**Comment by trivialfis** [comment]
I can work on https://github.com/conda-forge/xgboost-feedstock/issues/150#issuecomment-1820930122 after this PR.

File: CMakeLists.txt [hunk 1]
```diff
@@ -115,6 +119,9 @@ endif()
 if(BUILD_WITH_SHARED_NCCL AND (NOT USE_NCCL))
   message(SEND_ERROR "Build XGBoost with -DUSE_NCCL=ON to enable BUILD_WITH_SHARED_NCCL.")
 endif()
+if(USE_DLOPEN_NCCL AND (NOT USE_NCCL))
```
Comment (trivialfis) [inline CMakeLists.txt:122]: Add a check to disable Windows build.

File: plugin/federated/federated_comm.cuh [hunk 2]
```diff
@@ -16,5 +18,9 @@ class CUDAFederatedComm : public FederatedComm {
  public:
   explicit CUDAFederatedComm(Context const* ctx, std::shared_ptr<FederatedComm const> impl);
   [[nodiscard]] auto Stream() const { return stream_; }
+  Comm* MakeCUDAVar(Context const*, std::shared_ptr<Coll>) const override {
```
Comment (trivialfis) [inline plugin/federated/federated_comm.cuh:21]: Maybe define a CPU variant as base in the future.

File: tests/test_distributed/test_gpu_with_dask/test_gpu_with_dask.py [hunk 3]
```diff
@@ -572,6 +573,25 @@ def test_with_asyncio(local_cuda_client: Client) -> None:
     assert isinstance(output["history"], dict)
 
 
+def test_invalid_nccl(local_cuda_client: Client) -> None:
```
Comment (trivialfis) [inline tests/test_distributed/test_gpu_with_dask/test_gpu_with_dask.py:576]: Add a test to make sure NCCL is not loaded when distributed training is not used.

File: doc/tutorials/dask.rst [hunk 4]
```diff
@@ -536,6 +536,17 @@ Troubleshooting
...
   that includes `Multiple processes within a communication group ...` upon initialization.
 
+- Failed to load NCCL from the environment. Starting from 2.1.0, in order to reduce the
+  binary size, the XGBoost PyPI package loads NCCL from the environment instead of relies
+  on static linking. If you encounter an error message like "Failed to load nccl ...",
+  please install nccl using pip:
+
+.. code-block:: sh
+
+    pip install pip install nvidia-nccl-cu12
```
Comment (rongou) [inline doc/tutorials/dask.rst]: `pip install` repeated?

Comment (trivialfis) [reply doc/tutorials/dask.rst]: Thanks for catching that, fixed.

**Review by rongou — APPROVED** [review]

**Review by RAMitchell — APPROVED** [review]

**Pull request merged**